### ISSUE TITLE
AzCopy fails to download gzip-ed blob that has the ContentEncoding property set to gzip

### ISSUE THREAD

**Issue by dterziev** [Azure/azure-storage-azcopy#374]
### Which version of the AzCopy was used? 
``` cmd
> azcopy_windows_amd64.exe --version
azcopy version 10.1.0
```

### Which platform are you using? (ex: Windows, Mac, Linux)
Windows

##### Note: Please remove the SAS to avoid exposing your credentials. If you cannot remember the exact command, please retrieve it from the beginning of the log file.

This command is actually issued by Storage Explorer 1.8.0 when using the bundled version of AzCopy.

```
azcopy_windows_amd64.exe copy "source blob url with SAS" "local path" --overwrite=true --recursive --from-to=BlobLocal --check-md5 "FailIfDifferent"
```

### What problem was encountered?
Download fails the MD5 check with this error in the log file:

``` text
2019/05/08 21:21:47 ERR: [P#0-T#215] DOWNLOADFAILED: <url of the blob here>  : 000 : the MD5 hash of the data, as we received it, did not match the expected value, as found in the Blob/File Service. This means that either there is a data integrity error OR another tool has failed to keep the stored hash up to date. When Checking MD5 hash. X-Ms-Request-Id: 
```

### How can we reproduce the problem in the simplest way?
1. Upload a gzip-ed blob and set the ContentEncoding=gzip property and ContentMD5 to the hash of the compressed blob, this is a C# example:

``` c#
            CloudBlockBlob blob = (get blob reference);
            blob.Properties.ContentType = "application/xml";
            blob.Properties.ContentEncoding = "gzip";

            using (var outputStream = await blob.OpenWriteAsync().ConfigureAwait(false))
            using (var gzipStream = new GZipStream(outputStream, CompressionLevel.Optimal))
            {
                // write the bytes to gzipStream here
            }
```

2. Try downloading the blob from Storage Explorer (v1.8.0) with AzCopy and that will fail.

I suspect that the MD5 validation is performed after the content is decompressed because the HTTP response has the `Content-Encoding: gzip` header. This is, of course, different from the MD5 on the compressed blob.

### Have you found a mitigation/solution?

* Download with `--check-md5 "NoCheck"` - this skips the validation but then the blob is actualy decompressed which is a side-effect that may not be desirable.
* Remove the ContentEncoding property on the blob but then if a normal HTTP client downloads the content it does not decompress it which in my particular case leads to invalid XML file content.

It seems that the MD5 validation in AzCopy needs to be performed before the blob content is decompressed or better, not decompress the blob at all.

**Comment by JohnRusk** [comment]
Thanks for the detailed description @dterziev. We'll have to give this some thought.  Thanks for bringing it to our attention.

**Comment by JohnRusk** [comment]
I can reproduce it. (And, incidentally, if you have a very small XML file, such that the compressed size is actually larger than the uncompressed size, you get a different error even before the MD5 check happens.)

**Comment by JohnRusk** [comment]
@dterziev I've also found that the work around you propose, re `--check-md5 NoCheck` doesn't work.  Why? Because although the file is uncompressed, its size is truncated to the test size.  (At least it is in my test, on Windows Server 2016).

**Comment by JohnRusk** [comment]
My preferred solution will be that we do not decompress the blob at all. I've proposed that to the rest of the AzCopy team, and hope to have the change made promptly.

**Comment by JohnRusk** [comment]
We'll definitely change the code to not decompress the blob at all. (Decompression causes other problems too, including the fact that the file size on disk, of the downloaded file, is limited by AzCop's current implementation to that of the compressed file, not the decompressed content, so decompression actually loses data).

**Comment by adreed-msft** [comment]
Hi, this was resolved in AzCopy v10.1.2. Enjoy!
